### PR TITLE
Fix two memory leaks

### LINKED ISSUE #10338: Memory leak
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

When running prefect app in an infinite loop (as a job that should listen to e.g. some messaging service and process the events) we are running into infinite memory growth. We tried to inspect our code but even if we run the most minimal example without persisting any data and also garbage collecting the memory still grows.

### Reproduction

```python3
import gc
import asyncio
import time

from prefect import flow, task



@task(
    #retries=3,
    #persist_result=False,
    #cache_result_in_memory=False,
)
async def build_dataframe_async(id, index):
    print(f"Hello World = id = {id}, index = {index}")


@flow(
    # log_prints=True,
    # persist_result=False,
    cache_result_in_memory=False,
)
async def flow_async(id):
    for index in range(30):
        await build_dataframe_async(id, index)


if __name__ == '__main__':
    for index in range(5):
        print(f"____________________ Index = {index} ____________________")
        asyncio.run(flow_async(index))

        gc.collect(0)
        gc.collect(1)
        gc.collect(2)
        time.sleep(0.01)
        gc.collect()
        time.sleep(0.01)
```

### Error

_No response_

### Versions

```Text
python 3.10
prefect 2.10.18
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by abrookins**
Attempts to resolve two memory leaks:
1. `QueueService` instances keep a reference to their asyncio tasks even after shutdown. The problem with this is that the task's context contains a reference to the `EngineContext` we use for a flow run, and _that_ contains references to every task call made during the flow run. As we process runs, these objects accumulate in memory without getting released. This change deletes the asynctio task after the `QueueService` stops, which fixes the leak.
2. `EventLoopThread` instances keep a reference to callbacks that should fire when the thread shuts down. The problem here is that these thread instances stick around and hold onto the calls, which accumulate in memory. After the thread shuts down and we run the callbacks, we don't need references to them, so this PR clears the list of callbacks after shutdown.

Using example code from #10338 to reproduce the leaks and test this code, we get the following with Prefect 2.16.2:

2-16-2-leak-50-runs

And here's the same code running with this PR:
Latest-fix-leak-50-runs

Fixes #10338.

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `maintenance`, `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed.

For new functions or classes in the Python SDK:

- [ ] This pull request includes helpful docstrings.
- [ ] If a new Python file was added, this pull request contains a stub page in the Python SDK docs and an entry in `mkdocs.yml` navigation.

### CODE REVIEW

File: src/prefect/_internal/concurrency/services.py [hunk 1]
```diff
@@ -91,6 +91,11 @@ def _stop(self, at_exit: bool = False):
             self._remove_instance()
             self._stopped = True
 
+            # Allow asyncio Task to be garbage-collected because it may
+            # contain references to all Prefect Task calls during a flow
+            # run.
+            del self._task
```
Comment (abrookins) [inline src/prefect/_internal/concurrency/services.py:97]: I need to look at this again next week, not sure this is how we want to clean up the task.

**Review by desertaxle — APPROVED** [review]
🧼

**Review by bunchesofdonald — APPROVED** [review]
Gorgeous, thanks for tracking those down!

**Review by urimandujano — APPROVED** [review]
Great finds! 🧙🏼

**Review by chrisguidry — APPROVED** [review]
Amazing find, thank you!

**Pull request merged**